Source/Scene/Cesium3DTilesetTraversal.js

```diff
@@ -0,0 +1,825 @@
+            var childrenLength = children.length;
+
+            children.sort(sortChildrenByDistanceToCamera);
+
+            if (shouldSelect) {
+                if (tile.refine === Cesium3DTileRefine.ADD) {
+                    tile._finalResolution = true;
+                    selectTile(tileset, tile, frameState);
+                } else {
+                    tile._selectionDepth = ancestorStack.length;
+
+                    if (tile._selectionDepth > 0) {
+                        tileset._hasMixedContent = true;
+                    }
+
+                    lastAncestor = tile;
+
+                    if (childrenLength === 0) {
+                        tile._finalResolution = true;
+                        selectTile(tileset, tile, frameState);
+                        continue;
+                    }
+
+                    ancestorStack.push(tile);
+                    tile._stackLength = stack.length;
+                }
+            }
+
+            for (var i = 0; i < childrenLength; ++i) {
+                var child = children[i];
+                stack.push(child);
+            }
+        }
+    }
+
+    function selectTile(tileset, tile, frameState) {
+        // There may also be a tight box around just the tile's contents, e.g., for a city, we may be
+        // zoomed into a neighborhood and can cull the skyscrapers in the root node.
+        if (tile.contentAvailable && (
+                (tile._visibilityPlaneMask === CullingVolume.MASK_INSIDE) ||
+                (tile.contentVisibility(frameState) !== Intersect.OUTSIDE)
+            )) {
+            tileset._selectedTiles.push(tile);
+
+            var tileContent = tile.content;
+            if (tileContent.featurePropertiesDirty) {
+                // A feature's property in this tile changed, the tile needs to be re-styled.
+                tileContent.featurePropertiesDirty = false;
+                tile.lastStyleTime = 0; // Force applying the style to this tile
+                tileset._selectedTilesToStyle.push(tile);
+            }  else if ((tile._lastSelectedFrameNumber !== frameState.frameNumber - 1)) {
+                // Tile is newly selected; it is selected this frame, but was not selected last frame.
+                tileset._selectedTilesToStyle.push(tile);
+            }
+            tile._lastSelectedFrameNumber = frameState.frameNumber;
+        }
+    }
+
+    // PERFORMANCE_IDEA: is it worth exploiting frame-to-frame coherence in the sort, i.e., the
+    // list of children are probably fully or mostly sorted unless the camera moved significantly?
+    function sortChildrenByDistanceToCamera(a, b) {
+        // Sort by farthest child first since this is going on a stack
+        if (b._distanceToCamera === 0 && a._distanceToCamera === 0) {
+            return b._centerZDepth - a._centerZDepth;
+        }
+
+        return b._distanceToCamera - a._distanceToCamera;
+    }
+
+    var emptyArray = freezeObject([]);
```

Comment:
Several of the functions in here operate by returning an empty array, so this just avoids the allocation and guarantees that it is read only.